src/core/server/saved_objects/permission_control/acl.ts

```diff
@@ -0,0 +1,337 @@
+    if (!principals.groups) {
+      principals.groups = [];
+    }
+    principals.groups = Array.from(new Set([...principals.groups, ...groups]));
+  }
+  return principals;
+};
+
+const deleteFromPrincipals = ({
+  principals,
+  users,
+  groups,
+}: {
+  principals?: Principals;
+  users?: string[];
+  groups?: string[];
+}) => {
+  if (!principals) {
+    return principals;
+  }
+  if (users && principals.users) {
+    principals.users = principals.users.filter((item) => !users.includes(item));
+  }
+  if (groups && principals.groups) {
+    principals.groups = principals.groups.filter((item) => !groups.includes(item));
+  }
+  return principals;
+};
+
+const checkPermission = (
+  allowedPrincipals: Principals | undefined,
+  requestedPrincipals: Principals
+) => {
+  return (
+    (allowedPrincipals?.users &&
+      requestedPrincipals?.users &&
+      checkPermissionForSinglePrincipalType(allowedPrincipals.users, requestedPrincipals.users)) ||
+    (allowedPrincipals?.groups &&
+      requestedPrincipals?.groups &&
+      checkPermissionForSinglePrincipalType(allowedPrincipals.groups, requestedPrincipals.groups))
+  );
+};
+
+const checkPermissionForSinglePrincipalType = (
+  allowedPrincipalArray: string[],
+  requestedPrincipalArray: string[]
+) => {
+  return (
+    allowedPrincipalArray &&
+    requestedPrincipalArray &&
+    (allowedPrincipalArray.includes('*') ||
+      requestedPrincipalArray.some((item) => allowedPrincipalArray.includes(item)))
+  );
+};
+
+export class ACL {
+  private permissions?: Permissions;
+  constructor(initialPermissions?: Permissions) {
+    this.permissions = initialPermissions || {};
+  }
+
+  /**
+   * A function that parses the permissions object to check whether the specific principal has the specific permission types or not
+   *
+   * @param {Array} permissionTypes permission types
+   * @param {Object} principals the users or groups
+   * @returns true if the principal has the specified permission types, false if the principal has no permission
+   *
+   * @public
+   */
+  public hasPermission(permissionTypes: string[], principals: Principals) {
+    if (!permissionTypes || permissionTypes.length === 0 || !this.permissions || !principals) {
+      return false;
+    }
+
+    const currentPermissions = this.permissions;
+    return permissionTypes.some((permissionType) =>
+      checkPermission(currentPermissions[permissionType], principals)
+    );
+  }
+
+  /**
+   * A permissions object build function that adds principal with specific permission to the object
+   *
+   * This function is used to contruct a new permissions object or add principals with specified permissions to
+   * the existing permissions object. The usage is:
+   *
+   * const permissionObject = new ACL()
+   *  .addPermission(['write', 'management'], {
+   *     users: ['user2'],
+   *   })
+   *   .addPermission(['write', 'management'], {
+   *     groups: ['group1'],
+   *   })
+   *   .getPermissions();;
```

Comment:
Extra semicolon for `.getPermissions();;`

Reply:
Thanks for catching that, I've removed it.